ingestion-core/src/test/java/com/mozilla/telemetry/ingestion/core/transform/PubsubMessageToObjectNodeBeamTest.java

```diff
@@ -0,0 +1,539 @@
+          Field.of("key", LegacySQLTypeName.STRING), //
+          Field.of("value", LegacySQLTypeName.INTEGER)) //
+      .setMode(Mode.REPEATED).build();
+
+  private static final Field MAP_FIELD_WITHOUT_VALUE = Field //
+      .newBuilder("map_field", LegacySQLTypeName.RECORD, //
+          Field.of("key", LegacySQLTypeName.STRING)) //
+      .setMode(Mode.REPEATED).build();
+
+  private static final PubsubMessageToObjectNode.Payload TRANSFORM = //
+      PubsubMessageToObjectNode.Payload.of(null, null, null);
+
+  @Test
+  public void testConvertFieldNameForBq() {
+    assertEquals("snake_case", PubsubMessageToObjectNode.Payload.convertNameForBq("snake_case"));
+    assertEquals("camel_case", PubsubMessageToObjectNode.Payload.convertNameForBq("camelCase"));
+    assertEquals("ram", PubsubMessageToObjectNode.Payload.convertNameForBq("RAM"));
+    assertEquals("active_gm_plugins",
+        PubsubMessageToObjectNode.Payload.convertNameForBq("activeGMPlugins"));
+    assertEquals("d2d_enabled", PubsubMessageToObjectNode.Payload.convertNameForBq("D2DEnabled"));
+    assertEquals("gpu_active", PubsubMessageToObjectNode.Payload.convertNameForBq("GPUActive"));
+    assertEquals("product_id", PubsubMessageToObjectNode.Payload.convertNameForBq("ProductID"));
+    assertEquals("l3cache_kb", PubsubMessageToObjectNode.Payload.convertNameForBq("l3cacheKB"));
+    assertEquals("_64bit", PubsubMessageToObjectNode.Payload.convertNameForBq("64bit"));
+    assertEquals("hi_fi", PubsubMessageToObjectNode.Payload.convertNameForBq("hi-fi"));
+    assertEquals("untrusted_modules",
+        PubsubMessageToObjectNode.Payload.convertNameForBq("untrustedModules"));
+    assertEquals("xul_load_duration_ms",
+        PubsubMessageToObjectNode.Payload.convertNameForBq("xulLoadDurationMS"));
+    assertEquals("a11y_consumers",
+        PubsubMessageToObjectNode.Payload.convertNameForBq("A11Y_CONSUMERS"));
+  }
+
+  @Test
+  public void testCoerceNestedArrayToJsonString() throws Exception {
+    ObjectNode parent = Json.createObjectNode().set("events",
+        Json.createArrayNode().add(Json.createArrayNode().add("hi").add("there")));
+    List<Field> bqFields = ImmutableList
+        .of(Field.newBuilder("events", LegacySQLTypeName.STRING).setMode(Mode.REPEATED).build());
+    Map<String, Object> expected = ImmutableMap.of("events",
+        ImmutableList.of("[\"hi\",\"there\"]"));
+    TRANSFORM.transformForBqSchema(parent, bqFields, null);
+    assertEquals(expected, Json.asMap(parent));
+  }
+
+  @Test
+  public void testNullRecord() throws Exception {
+    ObjectNode parent = Json.createObjectNode().putNull("record");
+    ObjectNode additionalProperties = Json.createObjectNode();
+    List<Field> bqFields = ImmutableList.of(
+        Field.of("record", LegacySQLTypeName.RECORD, Field.of("field", LegacySQLTypeName.STRING)));
+    TRANSFORM.transformForBqSchema(parent, bqFields, additionalProperties);
+    assertEquals(ImmutableMap.of(), Json.asMap(parent));
```

Comment:
this is the only expected test result that changed. Beam would have previously encoded this as `"{"record":null}`, rather than `{}`.